File: 1001-1100/1059-All Paths from Source Lead to Destination/1059-All Paths from Source Lead to Destination.py
import collections
class Solution:
    def leadsToDestination(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        graph = collections.defaultdict(list)
        for u, v in edges:
            graph[u].append(v)
        
        visited = [0] * n
        def dfs(curr):
            if visited[curr] == 1:
                return True
            elif visited[curr] == 2:
                return False
            elif len(graph[curr]) == 0:
                return curr == destination
            else:
                visited[curr] = 2
                for neighbor in graph[curr]:
                    if not dfs(neighbor):
                        return False
                visited[curr] = 1
                return True
        return dfs(source)

File: 1001-1100/1059-All Paths from Source Lead to Destination/1059-All Paths from Source Lead to Destination.cpp
class Solution {
public:
    bool leadsToDestination(int n, vector<vector<int>>& edges, int source, int destination) {
        vector<vector<int>> graph(n);
        for (auto edge : edges) {
            graph[edge[0]].push_back(edge[1]);
        }
        
        vector<int> visited(n);
        return dfs(graph, source, visited, destination);
    }
    
private:
    bool dfs(vector<vector<int>>& graph, int curr, vector<int>& visited, int destination) {
        if (visited[curr] == 1) {
            return true;
        }
        else if (visited[curr] == 2) {
            return false;
        }
        else if (graph[curr].size() == 0) {
            return curr == destination;
        }
        else {
            visited[curr] = 2;
            for (int neighbor : graph[curr]) {
                if (!dfs(graph, neighbor, visited, destination)) {
                    return false;
                }
            }
            visited[curr] = 1;
            return true;
        }
    }
};
